File: implementation/csharp/GameEventScript.Conformance/fixture-exporter/NumberTextProbe.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System.Globalization;
using System.Text.Json;
using GameEventScript.Api;
using GameEventScript.Runtime.Values;

// Executable-only interoperability adapter. Parsing and formatting run through the VM.
internal static class NumberTextProbe
{
    internal static int Run(string input, string output)
    {
        using var requests = JsonDocument.Parse(File.ReadAllText(input));
        var program = GameEventScriptBuilder.Create().AddScript(
            "on Probe(value, text) { emit Result(formatted: value as :Text, number: text as :Number, percentage: text as :Percentage) }").Compile();
        var host = GameEventScriptHost.CreateBuilder().Build();
        var capture = new Capture();
        host.Subscribe("Result", ["formatted", "number", "percentage"], capture);
        host.Load(program);
        if (host.Start().State != GameEventScriptStartState.Ready) throw new InvalidOperationException("Probe host failed to start.");
        var responses = new List<object>();
        foreach (var row in requests.RootElement.EnumerateArray())
        {
            var kind = row.GetProperty("kind").GetString();
            var payload = row.GetProperty("payload").GetString()!;
            var unit = row.GetProperty("unit").GetString() switch
            {
                "none" => GameEventScriptBytecodeInstructionUnit.UnitNone,
                "m" => GameEventScriptBytecodeInstructionUnit.UnitMeter,
                "s" => GameEventScriptBytecodeInstructionUnit.UnitSecond,
                "degree" => GameEventScriptBytecodeInstructionUnit.UnitDegree,
                _ => throw new InvalidOperationException("Unknown unit.")
            };
            var value = kind == "integer" ? GesValue.GesInteger(long.Parse(payload, CultureInfo.InvariantCulture), unit)
                : DecodeFloat(kind!, payload, unit);
            var text = row.TryGetProperty("text", out var textInput) ? textInput.GetString()! : "nothing";
            capture.Message = null;
            capture.Count = 0;
            host.Receive(GameEventScriptMessage.Create("Probe", [new("value", value), new("text", GesValue.GesText(text))]));
            if (host.RunToCompletion().State != GameEventScriptExecutionState.Completed || capture.Count != 1 || capture.Message is not { } result)
                throw new InvalidOperationException("Probe did not produce exactly one result.");
            responses.Add(new { source = Describe(value), text = result.Arguments[0].AsText(), number = Describe(result.Arguments[1]), percentage = Describe(result.Arguments[2]) });
        }
        File.WriteAllText(output, JsonSerializer.Serialize(responses));
        return 0;
    }

    private static GesValue DecodeFloat(string kind, string payload, GameEventScriptBytecodeInstructionUnit unit)
    {
        var value = BitConverter.Int64BitsToDouble(unchecked((long)ulong.Parse(payload, NumberStyles.HexNumber, CultureInfo.InvariantCulture)));
        return kind switch
        {
            "float" => GesValue.GesFloat(value, unit),
            "percentage" when unit == GameEventScriptBytecodeInstructionUnit.UnitNone => GesValue.GesPercentage(value),
            _ => throw new InvalidOperationException("Unknown numeric input kind.")
        };
    }

    private static string Describe(GesValue value)
    {
        if (value.IsNothing) return "nothing";
        var unit = value.ValueUnit switch
        {
            GameEventScriptBytecodeInstructionUnit.UnitNone => "none",
            GameEventScriptBytecodeInstructionUnit.UnitMeter => "m",
            GameEventScriptBytecodeInstructionUnit.UnitSecond => "s",
            GameEventScriptBytecodeInstructionUnit.UnitDegree => "degree",
            _ => throw new InvalidOperationException("Unexpected result unit.")
        };
        if (value.ValueKind == GameEventScriptBytecodeTypeKind.Integer) return "integer:" + value.AsInteger().ToString(CultureInfo.InvariantCulture) + ":" + unit;
        var kind = value.ValueKind == GameEventScriptBytecodeTypeKind.Percentage ? "percentage" : value.ValueKind == GameEventScriptBytecodeTypeKind.Float ? "float" : throw new InvalidOperationException("Non-numeric result.");
        return kind + ":" + unchecked((ulong)BitConverter.DoubleToInt64Bits(value.AsNumber())).ToString("x16", CultureInfo.InvariantCulture) + ":" + unit;
    }

    private sealed class Capture : IGameEventScriptNativeMessageHandler
    {
        internal GameEventScriptMessage? Message;
        internal int Count;

        /// <inheritdoc />
        public void Handle(GameEventScriptMessage message, GameEventScriptContext context)
        {
            Message = message;
            Count++;
        }
    }
}

File: implementation/swift/GameEventScriptConformance/Sources/GameEventScriptConformanceTool/NumberTextProbe.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import Foundation
import GameEventScriptCompiler
import GameEventScriptRuntime

// Executable-only interoperability adapter. Parsing and formatting run through the VM.
enum NumberTextProbe {
    private struct Request: Decodable {
        let kind: String
        let payload: String
        let unit: String
        let text: String?
    }
    private struct Response: Encodable {
        let source: String
        let text: String
        let number: String
        let percentage: String
    }
    private final class Capture: GameEventScriptNativeMessageHandler {
        var message: GameEventScriptMessage?
        var count = 0
        func handle(_ message: GameEventScriptMessage, context: GameEventScriptContext) {
            self.message = message
            count += 1
        }
    }
    static func run(input: String, output: String) throws {
        let requests = try JSONDecoder().decode([Request].self, from: Data(contentsOf: URL(fileURLWithPath: input)))
        let program = try GameEventScriptBuilder().addScript(
            "on Probe(value, text) { emit Result(formatted: value as :Text, number: text as :Number, percentage: text as :Percentage) }"
        ).compile()
        let host = try GameEventScriptHost(seed: 0)
        let capture = Capture()
        _ = try host.subscribe(
            .init(name: "Result", parameters: ["formatted", "number", "percentage"]), handler: capture)
        _ = try host.load(program)
        guard try host.start().state == .ready else { throw ToolError.invalidFixture("Probe host failed to start") }
        var responses: [Response] = []
        for row in requests {
            let unit: GesUnit
            switch row.unit {
            case "none": unit = .none
            case "m": unit = .meter
            case "s": unit = .second
            case "degree": unit = .degree
            default: throw ToolError.invalidFixture("Unknown unit")
            }
            let value: GesValue
            if row.kind == "integer", let integer = Int64(row.payload) {
                value = .integer(integer, unit: unit)
            } else if let bits = UInt64(row.payload, radix: 16) {
                let number = Double(bitPattern: bits)
                if row.kind == "float" {
                    value = .float(number, unit: unit)
                } else if row.kind == "percentage" && unit == .none {
                    value = .percentage(number)
                } else {
                    throw ToolError.invalidFixture("Unknown numeric input kind")
                }
            } else {
                throw ToolError.invalidFixture("Invalid numeric payload")
            }
            capture.message = nil
            capture.count = 0
            host.receive(
                try .init(
                    name: "Probe",
                    arguments: [
                        .init(name: "value", value: value), .init(name: "text", value: .text(row.text ?? "nothing")),
                    ]))
            guard try host.runToCompletion().state == .completed, capture.count == 1, let result = capture.message
            else {
                throw ToolError.invalidFixture("Probe did not produce exactly one result")
            }
            responses.append(
                try Response(
                    source: describe(value), text: result.arguments[0].asText,
                    number: describe(result.arguments[1]), percentage: describe(result.arguments[2])))
        }
        try JSONEncoder().encode(responses).write(to: URL(fileURLWithPath: output))
    }
    private static func describe(_ value: GesValue) throws -> String {
        if value.isNothing { return "nothing" }
        let unit: String
        switch value.unit {
        case .none: unit = "none"
        case .meter: unit = "m"
        case .second: unit = "s"
        case .degree: unit = "degree"
        }
        if let integer = value.integerValue { return "integer:\(integer):\(unit)" }
        guard value.kind == .float || value.kind == .percentage else {
            throw ToolError.invalidFixture("Non-numeric result")
        }
        let kind = value.kind == .percentage ? "percentage" : "float"
        let hex = String(value.asNumber.bitPattern, radix: 16)
        return kind + ":" + String(repeating: "0", count: 16 - hex.count) + hex + ":" + unit
    }
}
